src/rebar_app_discover.erl

```diff
@@ -333,27 +335,61 @@ find_app(AppInfo, AppDir, SrcDirs, Validate, State) ->
-    Config = rebar_config:consult(AppDir),
-    AppInfo1 = rebar_app_info:update_opts(AppInfo, rebar_app_info:opts(AppInfo), Config),
-    find_app_(AppInfo1, AppDir, SrcDirs, Validate).
+    find_app_(AppInfo1, AppDir, SrcDirs, Validate, State).
 
 -spec find_app_(rebar_app_info:t(), file:filename_all(),
-               [file:filename_all()], valid | invalid | all) ->
+               [file:filename_all()], valid | invalid | all,  rebar_state:t()) ->
     {true, rebar_app_info:t()} | false.
-find_app_(AppInfo, AppDir, SrcDirs, Validate) ->
-    AppFile = filelib:wildcard(filename:join([AppDir, "ebin", "*.app"])),
-    AppSrcFile = lists:append(
-        [filelib:wildcard(filename:join([AppDir, SrcDir, "*.app.src"]))
-         || SrcDir <- SrcDirs]
-    ),
-    AppSrcScriptFile = lists:append(
-        [filelib:wildcard(filename:join([AppDir, SrcDir, "*.app.src.script"]))
-         || SrcDir <- SrcDirs]
-    ),
-    try_handle_app_file(AppInfo, AppFile, AppDir, AppSrcFile, AppSrcScriptFile, Validate).
+find_app_(AppInfo, AppDir, SrcDirs, Validate, State) ->
+    Extensions = rebar_state:get(State, application_resource_extensions, ?DEFAULT_APP_RESOURCE_EXT),
+    ResourceFiles = [
+        {app, filelib:wildcard(filename:join([AppDir, "ebin", "*.app"]))} |
+        [
+        {extension_type(Ext), lists:append([ filelib:wildcard(filename:join([AppDir, SrcDir, "*." ++ Ext]))
+        || SrcDir <- SrcDirs])}
+        || Ext <- Extensions
+    ]],
+    % io:format(user, "conf: ~p~n", [{AppDir, SrcDirs}]),
+    % io:format(user, "Extensions: ~p~n", [Extensions]),
+    % io:format(user, "ResourceFiles: ~p~n", [ResourceFiles]),
+    FlattenedResourceFiles = flatten_resource_files(ResourceFiles),
+    % io:format(user, "FlattenedResourceFiles: ~p~n", [FlattenedResourceFiles]),
+    try_handle_resource_files(AppInfo, AppDir, FlattenedResourceFiles, Validate).
+
+-spec extension_type(string()) -> app_resource_type().
+extension_type(Extension) ->
+    Mapping = [
+        {"app", app},
+        {"src", app_src},
+        {"script", script},
+        {"mix.exs", mix_exs}
+    ],
+    extension_type(Mapping, Extension).
+
+-spec extension_type([{string(), Type}], string()) -> Type
+    when Type :: app_resource_type().
+extension_type([], _) ->
+    %% default to app_src
+    app_src;
+extension_type([{Pattern, Type} | Rest], Extension) ->
+    case lists:suffix(Pattern, Extension) of
+        true ->
+            Type;
+        false ->
+            extension_type(Rest, Extension)
+    end.
+
+-spec flatten_resource_files(ResourceFiles) -> FlattenedResourceFiles
+    when ResourceFiles :: [{app_resource_type(), [file:filename()]}],
+         FlattenedResourceFiles :: [{app_resource_type(), file:filename()}].
+flatten_resource_files(ResourceFiles) ->
+    lists:foldr(
+        fun ({_Type, []}, Acc)      -> Acc;
+            ({Type, [File]}, Acc)   -> [{Type, File} | Acc];
+            ({_Type, Others}, _Acc) -> throw({error, {multiple_app_files, Others}})
```

Comment:
Are you likely to hit this if you use profile-specific extensions? You'd have to only use `.<profile>.app.src` to have many without them clashing, and the moment we have an older rebar3 version reading them, it should complain of multiple app files or not finding a valid one, right? So you'd want to leave in a default `.app.src` for compatibility and then this will start complaining.

The end result is that a backwards-compatible profile-specific things like `.app.<profile>.src` would be the emergent pattern. I think that's valid if you go down that road, but yeah. Surprising a bit.

Reply:
Ah yes, let's have files only appear in one of the buckets according to the defined priorities.